Python/Product/PythonTools/PythonTools/Repl/InteractiveWindowProvider.cs

```diff
@@ -30,202 +34,178 @@ namespace Microsoft.PythonTools.Repl {
-                        .Select(r => r.Name)
-                        .ToArray();
-                    return evaluator;
+            window.InteractiveWindow.TextView.Closed += (s, e) => {
+                lock (_windows) {
+                    Debug.Assert(ReferenceEquals(_windows[curId], window));
+                    _windows.Remove(curId);
+                    _temporaryWindows.Remove(replId);
                 }
-            }
-            return null;
-        }
+            };
 
-        private InteractiveWindowInfo CreateInteractiveWindow(IInteractiveEvaluator/*!*/ evaluator, IContentType/*!*/ contentType, string[] roles, int id, string/*!*/ title, Guid languageServiceGuid, string replId) {
-            return CreateInteractiveWindowInternal(evaluator, contentType, roles, id, title, languageServiceGuid, replId);
+            return window;
         }
 
-        private InteractiveWindowInfo CreateInteractiveWindowInternal(IInteractiveEvaluator evaluator, IContentType contentType, string[] roles, int id, string title, Guid languageServiceGuid, string replId) {
-            var service = (IVsUIShell)ServiceProvider.GlobalProvider.GetService(typeof(SVsUIShell));
-            var model = (IComponentModel)ServiceProvider.GlobalProvider.GetService(typeof(SComponentModel));
+        private IVsInteractiveWindow CreateInteractiveWindowInternal(
+            IInteractiveEvaluator evaluator,
+            IContentType contentType,
+            bool alwaysCreate,
+            int id,
+            string title,
+            Guid languageServiceGuid,
+            string replId
+        ) {
+            var service = (IVsUIShell)_serviceProvider.GetService(typeof(SVsUIShell));
+            var model = (IComponentModel)_serviceProvider.GetService(typeof(SComponentModel));
+
+            //SaveInteractiveInfo(id, evaluator, contentType, roles, title, languageServiceGuid, replId);
 
-            SaveInteractiveInfo(id, evaluator, contentType, roles, title, languageServiceGuid, replId);
+            var creationFlags =
+                __VSCREATETOOLWIN.CTW_fMultiInstance |
+                __VSCREATETOOLWIN.CTW_fActivateWithProject;
 
-            // we don't pass __VSCREATETOOLWIN.CTW_fMultiInstance because multi instance panes are
-            // destroyed when closed.  We are really multi instance but we don't want to be closed.  This
-            // seems to work fine.
-            __VSCREATETOOLWIN creationFlags = 0;
-            if (!roles.Contains("DontPersist")) {
+            if (alwaysCreate) {
                 creationFlags |= __VSCREATETOOLWIN.CTW_fForceCreate;
             }
 
             var replWindow = _windowFactory.Create(GuidList.guidPythonInteractiveWindowGuid, id, title, evaluator, creationFlags);
+            ((ToolWindowPane)replWindow).BitmapImageMoniker = KnownMonikers.PYInteractiveWindow;
+            replWindow.InteractiveWindow.Properties[typeof(IVsInteractiveWindow)] = replWindow;
```

Comment:
Why are we assigning this property?  It seems very strange that we're storing a property under an interface type that we don't own, and I don't see any other cases of typeof(IVsInteractiveWindow) being used as far as searching on github is concerned